packages/mediapipe-core/lib/src/io/task_executor.dart

```diff
@@ -0,0 +1,106 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ffi';
+import 'package:mediapipe_core/io.dart';
+import 'package:logging/logging.dart';
+
+final _log = Logger('TaskExecutor');
+
+/// {@template TaskExecutor}
+/// Instantiates and manages an object which can complete MediaPipe tasks. The
+/// managed task-completing object does not exist in Dart memory, but instead in
+/// platform-dependent native memory.
+///
+/// Extending classes should implement [createResultsPointer], [createWorker],
+/// [closeWorker], and any additional task-specific methods. Applications will
+/// only call those extra task-specific methods, which will throw an exception
+/// on any error communicating with the native workers. It is the job of
+/// application code to surround those task-specific methods with try-catch
+/// clauses.
+/// {@endtemplate}
+abstract class TaskExecutor<
+    NativeOptions extends Struct,
```

Comment:
> Can you explain to me why we would like to expose the bindings.Struct types here?

This connects to my attempt to have typed native structs for objects that originate in Dart code. An example is the `TaskOptions` subclasses that a Flutter application will instantiate to pass down to MediaPipe. My goal is to have those interactions typed all the way down to the ffigen-created bindings.

Doing this implies typed signatures of methods like `createWorker` and `createResultsPointer` in classes that extend `TaskExecutor`. It is potentially true that if those methods accepted `Pointer<void>` and cast their parameters locally, these types could be done away with. But, we know what the type of the pointers are, so at some point we would have to pretend we know less than we do, lest the analyzer would complain to the tune of 

```
The argument type 'Pointer<TextClassifierOptions>' can't be assigned to the parameter type 'Pointer<Void>'. 
```

> I was hoping that we would encapsulate the native types inside Dart types

Eventually, the bindings will accept a pointer to a typed struct. If I'm understanding everything correctly, we can acknowledge what that type is through the whole call chain, or treat it as a `Pointer<void>` until the last second when we cast it. Is that understanding correct?

> Also, because dart:ffi in general does not allow polymorphism on NativeTypes, you won't be able to do things such as calloc<NativeOptions>(), which means you'll make sub classes for every concrete type flowing into NativeOptions.

Subclasses for every concrete type is already happening, because each `TaskExecutor` class implements its own fully unique, task-specific methods with varying signatures. For example, `TextClassifierExecutor` will implement `classify(String text)`, whereas other executors will implement other functions (sometimes plural) with different signatures that cannot be easily standardized in a parent class.

I had hoped to avoid subclasses for every type, but I believe this detail would make it challenging and convoluted to do so.

Reply:
I agree that if we're giving users `Pointer`s, we should give them the correct type arguments and not `Pointer<Void>`. (Though what I had hoped was that we never give users `Pointer`s only classes that internally hold a pointer.)

> I had hoped to avoid subclasses for every type, but I believe this detail would make it challenging and convoluted to do so.

That doesn't really work when wrapping FFI code, because most of the FFI code cannot be polymorphic.

If you have a lot of repetitive code, consider writing a code generator! 🙈

Reply:
To clarify, the `TaskExecutor` subclasses are all implementation details that an app developer will never know about. They are completely concealed from actual Flutter code.